### PR TITLE
add $path column to the IcebergConnector

### PR DESCRIPTION
**Pull request by osscm**
## Description

Support `$path` column to the IcebergConnector.

ToDo: will be adding test-cases, and will do some of the refactoring.
Started so that can get early feedback.

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links
https://github.com/trinodb/trino/issues/8769

## Documentation

( ) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -302,6 +304,8 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
                     .filter(column -> partitionSourceIds.contains(column.getId()))
                     .collect(toImmutableMap(IcebergColumnHandle::getId, Function.identity()));
 
+            // add a column for the path.
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:307]: left over?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -375,7 +379,10 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
     {
         IcebergTableHandle table = (IcebergTableHandle) tableHandle;
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
-        return getColumns(icebergTable.schema(), typeManager).stream()
+        List<IcebergColumnHandle> columns = new ArrayList<>(getColumns(icebergTable.schema(), typeManager));
+        // add hidden path column
+        columns.add(getPathColumnHandle(columns.get(columns.size()-1).getId() + 1));
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:384]: I have some doubts regarding the association  of this column id on the fly. Why not using a statically defined negative value for $path, and other syntethic columns that may appear along the way?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 3]
```diff
@@ -375,7 +379,10 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
     {
         IcebergTableHandle table = (IcebergTableHandle) tableHandle;
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
-        return getColumns(icebergTable.schema(), typeManager).stream()
+        List<IcebergColumnHandle> columns = new ArrayList<>(getColumns(icebergTable.schema(), typeManager));
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:382]: Use ImmutableList.Builder instead.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 4]
```diff
@@ -375,7 +379,10 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
...
+        // add hidden path column
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:383]: the next line is self-explanatory making this comment rather redundant.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 5]
```diff
@@ -170,6 +172,24 @@ public static long resolveSnapshotId(Table table, long snapshotId)
                 .collect(toImmutableList());
     }
 
+    public static IcebergColumnHandle getPathColumnHandle(int id)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:175]: this belongs rather in IcebergColumnHandle class.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java [hunk 6]
```diff
@@ -170,6 +172,24 @@ public static long resolveSnapshotId(Table table, long snapshotId)
...
+    {
+        return new IcebergColumnHandle(
+                new ColumnIdentity(id, IcebergColumnHandle.PATH_COLUMN_NAME, PRIMITIVE, ImmutableList.of()),
+                VarcharType.createUnboundedVarcharType(),
+                ImmutableList.of(),
+                VarcharType.createUnboundedVarcharType(),
+                Optional.empty());
+    }
+
+    public static ColumnMetadata getPathColumnMetadata()
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:185]: can be placed in IcebergMetadata as private static

**Review by findinpath — CHANGES_REQUESTED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java (hunk 5 above)
Comment (osscm) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java:175]: thought about that, as done in the datalake connector case, but I felt keeping in `IcebergUtil` will make keep the `IcebergColumnHandle` cleaner and dont have anything `util` kind of function.

**Pull request closed without merging**